jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ChunkAccumulator.java

```diff
@@ -0,0 +1,228 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.jetty.io.Content.Chunk;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.CompletableTask;
+
+/**
+ * An accumulator of {@link Content.Chunk}s used to facilitate minimal copy
+ * aggregation of multiple chunks.
+ */
+public class ChunkAccumulator
+{
+    private static final ByteBufferPool NON_POOLING = new ByteBufferPool.NonPooling();
+    private final List<Chunk> _chunks = new ArrayList<>();
+    private int _size;
+
+    public ChunkAccumulator()
+    {
+    }
+
+    /**
+     * Add a {@link Chunk} to the accumulator.
+     * @param chunk The {@link Chunk} to accumulate.  If a reference is kept to the chunk (rather than a copy), it will be retained.
+     * @return true if the {@link Chunk} had content and was added to the accumulator.
+     */
+    public boolean add(Chunk chunk)
+    {
+        if (chunk.hasRemaining())
+        {
+            _size = Math.addExact(_size, chunk.remaining());
+            if (chunk.canRetain())
+            {
+                chunk.retain();
+                return _chunks.add(chunk);
+            }
+            return _chunks.add(Chunk.from(BufferUtil.copy(chunk.getByteBuffer()), chunk.isLast(), () -> {}));
+        }
+        return false;
+    }
+
+    /**
+     * Get the total size of the accumulated {@link Chunk}s.
+     * @return The total size in bytes.
+     */
+    public int size()
```

Comment:
I think this is better named `length()` as `size()` is unclear if it refers to the number of chunks or something else.
Or perhaps more explicitly `bytesLength()` or similar.